File: MechatronicsLab manuals/Lab5/ExampleCodes/WifiExample/toggleLED.m
function toggleLED
    % toggleLED.mlx
    % ---------------------------------------------------------------------
    % Example MATLAB script to connect to the Arduino Nano ESP32 (Alvik
    % robot) over TCP, send LED toggles with \r\n, and read incoming data
    % in the background using a callback.
    %
    % This version uses "configureTerminator" and "configureCallback"
    % to handle the reading of data line-by-line in MATLAB versions where
    % direct property assignment (e.g., t.Terminator = 'LF') is not allowed.
    %
    % REQUIREMENTS:
    % - MATLAB R2019b or later, but some versions only allow setting
    %   terminator/callback via these functions.
    % - Arduino Nano ESP32 running the non-blocking server code.

    % Clear workspace and command window (optional):
    clearvars; clc;

    % 1) Define the IP address and port of your Arduino Nano ESP32:
    arduinoIP   = '192.168.0.101';  % <-- Replace if needed
    arduinoPort = 5050;               % <-- Must match TCP_PORT in Arduino code

    % 2) Create a tcpclient object to connect to the Arduino.
    %    'Timeout' is how long MATLAB will wait for a response.
    try
        t = tcpclient(arduinoIP, arduinoPort, 'Timeout', 5);
        disp('Connected to Arduino Nano ESP32 via TCP.');
    catch ME
        warning('Failed to connect: %s', ME.message);
        return;
    end

    % 3) Configure the tcpclient object to read data in the background:
    %    We'll treat '\n' (LF) as the line terminator. Each time a full line
    %    is received, the callback "onBytesAvailable" is triggered.

    % -- Set the terminator to LF (so it triggers on '\n'):
    configureTerminator(t, "LF");

    % -- Configure the callback to fire upon "terminator":
    configureCallback(t, "terminator", @onBytesAvailable);

    % 4) Example toggling loop:
    %    We send "1\r\n" to turn the LED on, then "0\r\n" to turn it off, each second.
    numToggles = 10;  % number of toggles (on+off steps)
    for i = 1:numToggles
        % Turn LEDs on (green)
        write(t, "1", "string");  % Now we send "1\r\n"
        fprintf('Sent: 1 (LED ON)\n');
        pause(2);
    
        % Turn LEDs off
        write(t, "0", "string");  % Now we send "0\r\n"
        fprintf('Sent: 0 (LED OFF)\n');
        pause(2);
    end

    % 5) (Optional) read any leftover data in the buffer:
    if t.NumBytesAvailable > 0
        leftoverData = read(t, t.NumBytesAvailable, "uint8");
        disp("Leftover data from Arduino: " + string(char(leftoverData)));
    end

    % 6) Close the connection (tcpclient object is closed when cleared)
    clear t
    disp('Connection closed.');
end



function onBytesAvailable(src, ~)
    % onBytesAvailable is the callback that runs in the background
    % whenever a full line of data (per 'Terminator' = LF) arrives.
    %
    % NOTE: The second argument (~) is the event data, which we don't need here.

    % Read all available bytes as a string
    data = read(src, src.BytesAvailable, "uint8");
    strData = string(char(data));

    % If multiple lines arrived at once, split them by newline:
    lines = strsplit(strData, newline);

    for i = 1:length(lines)
        thisLine = strtrim(lines{i});
        if ~isempty(thisLine)
            fprintf("Received [BG]: %s\n", thisLine);
        end
    end
end

File: MechatronicsLab manuals/Lab5/ExampleCodes/WifiExample/toggleLED.py
#!/usr/bin/env python3

import socket
import threading
import time

def read_lines_in_background(sock, stop_event):
    """
    Continuously read from the socket in a background thread.
    Accumulate data in a buffer and split on newline characters.
    """
    buffer = b""
    while not stop_event.is_set():
        try:
            # Try reading some bytes (non-blocking or with a short timeout).
            data = sock.recv(1024)
            if not data:
                print("Connection closed by the Arduino.")
                stop_event.set()
                break
            
            # Append to the buffer
            buffer += data
            
            # Split on newline; handle each complete line
            while b"\n" in buffer:
                line, buffer = buffer.split(b"\n", 1)
                # Remove any trailing carriage returns/spaces
                line = line.replace(b"\r", b"").strip()
                if line:
                    print("Received [BG]:", line.decode("utf-8"))

        except socket.timeout:
            # If you set a timeout and no data arrives, we just loop again
            pass
        except OSError as e:
            # Socket closed or error
            print("Socket error:", e)
            stop_event.set()
            break


def main():
    # 1) Define your Arduino Nano ESP32's IP/Port
    arduino_ip = "192.168.0.101"  # <-- Replace with your device's IP
    arduino_port = 5050            # <-- Must match the TCP_PORT in your Arduino code

    # 2) Create and configure the socket
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    # Optional: set a short timeout so 'recv' won't block forever
    s.settimeout(2.0)

    # 3) Attempt to connect
    try:
        s.connect((arduino_ip, arduino_port))
        print(f"Connected to Arduino Nano ESP32 at {arduino_ip}:{arduino_port}")
    except Exception as e:
        print("Failed to connect:", e)
        return

    # 4) Start a background thread to read incoming data
    stop_event = threading.Event()
    listener_thread = threading.Thread(
        target=read_lines_in_background,
        args=(s, stop_event),
        daemon=True
    )
    listener_thread.start()

    # 5) Example toggling loop:
    #    Send "1\r\n" to turn the LED on, then "0\r\n" to turn it off.
    num_toggles = 10
    for i in range(num_toggles):
        # Turn LEDs on (green)
        s.sendall(b"1\r\n")
        print("Sent: 1 (LED ON)")
        time.sleep(2)

        # Turn LEDs off
        s.sendall(b"0\r\n")
        print("Sent: 0 (LED OFF)")
        time.sleep(2)

    # 6) Cleanup: signal the background thread to stop, then close the socket
    stop_event.set()
    listener_thread.join()
    s.close()
    print("Connection closed.")


if __name__ == "__main__":
    main()
